cpp/include/raft/distance/detail/kernels/gram_matrix.cuh

```diff
@@ -182,37 +312,257 @@ class GramMatrixBase {
+              dense_input_matrix_view_t<math_t> x2,
+              dense_output_matrix_view_t<math_t> out,
+              raft::device_resources const& handle)
   {
-    raft::distance::distance<raft::distance::DistanceType::L2Unexpanded, math_t, math_t, math_t>(
-      raft::device_resources(stream), x1, x2, out, n1, n2, n_cols, is_row_major);
+    // check is_row_major consistency
+    bool is_row_major = get_is_row_major(out);
+    ASSERT(is_row_major ? (x2.stride(1) == 1) : (x2.stride(0) == 1),
+           "GramMatrix leading dimensions for x2 and out do not match");
+
+    // check dimensions
+    auto x1_structure = x1.get_structure();
+    ASSERT(x1_structure.get_n_rows() == out.extent(0),
+           "GramMatrix input matrix dimensions for x1 and out do not match");
+    ASSERT(x2.extent(0) == out.extent(1),
+           "GramMatrix input matrix dimensions for x2 and out do not match");
+    ASSERT(x2.extent(1) == x1_structure.get_n_cols(),
+           "GramMatrix input matrix dimensions for x1 and x2 do not match");
+
+    // extract major stride
+    int ld2    = is_row_major ? x2.stride(0) : x2.stride(1);
+    int ld_out = is_row_major ? out.stride(0) : out.stride(1);
+
+    math_t alpha = 1.0;
+    math_t beta  = 0.0;
+
+    cusparseSpMatDescr_t descrX1;
+    RAFT_CUSPARSE_TRY(
+      raft::sparse::detail::cusparsecreatecsr(&descrX1,
+                                              x1_structure.get_n_rows(),
+                                              x1_structure.get_n_cols(),
+                                              x1_structure.get_nnz(),
+                                              const_cast<int*>(x1_structure.get_indptr().data()),
+                                              const_cast<int*>(x1_structure.get_indices().data()),
+                                              const_cast<math_t*>(x1.get_elements().data())));
+
+    auto order = is_row_major ? CUSPARSE_ORDER_ROW : CUSPARSE_ORDER_COL;
+
+    cusparseDnMatDescr_t descrX2;
+    RAFT_CUSPARSE_TRY(raft::sparse::detail::cusparsecreatednmat(
+      &descrX2, x2.extent(0), x2.extent(1), ld2, const_cast<math_t*>(x2.data_handle()), order));
+
+    cusparseDnMatDescr_t descrOut;
+    RAFT_CUSPARSE_TRY(
+      raft::sparse::detail::cusparsecreatednmat(&descrOut,
+                                                out.extent(0),
+                                                out.extent(1),
+                                                ld_out,
+                                                const_cast<math_t*>(out.data_handle()),
+                                                order));
+
+    auto alg = order == CUSPARSE_ORDER_COL ? CUSPARSE_SPMM_CSR_ALG1 : CUSPARSE_SPMM_CSR_ALG2;
+
+    // compute X1*X2^T
+    auto opX1 = CUSPARSE_OPERATION_NON_TRANSPOSE;
+    auto opX2 = CUSPARSE_OPERATION_TRANSPOSE;
+
+    size_t bufferSize;
+    RAFT_CUSPARSE_TRY(raft::sparse::detail::cusparsespmm_bufferSize(handle.get_cusparse_handle(),
```

Comment:
We should be exposing these operations directly through our public APIs that can accept our sparse types as arguments. Can you please move this logic to a `raft/sparse/linalg/spmm.cuh` primitive like we've done for `raft/linalg/gemm.cuh`? It could just accept pointers for now. It's on our roadmap to scrape through the sparse APIs and accept our new types everywhere.